### PR TITLE
Mvdb/backend as uri method

### LINKED ISSUE #3045: Unrecoverable error: ValueError invalid literal for int() with base 10: '53473,ds053473-a1.{snip}.mongolab.com'
This breaks mongo backend, because of https://github.com/celery/celery/commit/33f1dc829223da5499e05e271a6d49edde7ffc68. I believe `maybe_sanitize_url` doesn't play well with replica sets...

Full backtrace:
```
ValueError: invalid literal for int() with base 10: '53473,{snip}.mongolab.com'
  File "celery/worker/__init__.py", line 206, in start
    self.blueprint.start(self)
  File "celery/bootsteps.py", line 119, in start
    self.on_start()
  File "celery/apps/worker.py", line 170, in on_start
    string(self.colored.cyan(' \n', self.startup_info())),
  File "celery/apps/worker.py", line 232, in startup_info
    self.app.conf.CELERY_RESULT_BACKEND or 'disabled',
  File "kombu/utils/url.py", line 63, in maybe_sanitize_url
    return sanitize_url(url, mask)
  File "kombu/utils/url.py", line 58, in sanitize_url
    return as_url(*_parse_url(url), sanitize=True, mask=mask)
  File "kombu/utils/url.py", line 24, in _parse_url
    return (scheme, unquote(parts.hostname or '') or None, parts.port,
  File "python2.7/urlparse.py", line 113, in port
    port = int(port, 10)
```

### PR DESCRIPTION
**Pull request by m-vdb**
fixes #3045 #3068

### CODE REVIEW

File: celery/backends/cache.py [hunk 1]
```diff
@@ -149,3 +149,12 @@ def __reduce__(self, args=(), kwargs={}):
                  expires=self.expires,
                  options=self.options))
         return super(CacheBackend, self).__reduce__(args, kwargs)
+
+    def as_uri(self, *args, **kwargs):
+        """
+        Return the backend as an URI. It properly handles the
+        case of multiple servers. It doesn't try to sanitize
+        password because memcached URIs doesn't support them.
```
Comment (m-vdb) [inline celery/backends/cache.py]: I wasn't sure about that, LMK if I'm completely off

File: celery/backends/base.py [hunk 2]
```diff
@@ -678,5 +686,9 @@ def _is_disabled(self, *args, **kwargs):
         raise NotImplementedError(
             'No result backend configured.  '
             'Please see the documentation for more information.')
+
+    def as_uri(self, *args, **kwargs):
+        return 'disabled'
```
Comment (ask) [inline celery/backends/base.py]: Should be `disabled://` ?

File: celery/backends/cache.py (hunk 1 above)
Comment (ask) [inline celery/backends/cache.py]: Yeah, I think you can technically using SASL but the Celery cache backend doesn't let you do so via the URL

File: celery/backends/base.py (hunk 2 above)
Comment (m-vdb) [inline celery/backends/base.py]: I kept the [previous behaviour](https://github.com/celery/celery/pull/3079/files#diff-5057f62807f26f9a701a0d9bf9a2a2b3L209) but makes more sense yes

File: celery/backends/cache.py (hunk 1 above)
Comment (m-vdb) [inline celery/backends/cache.py]: that's what I figured :white_check_mark:

**Comment by m-vdb** [comment]
@ask @thedrow looks like this build is gonna pass. LMK what you think :smile:

**Comment by m-vdb** [comment]
build is failing on appveyor because of missing `pymongo`, strange the issue wasn't there on travis

**Comment by ask** [comment]
@m-vdb Oh yeah, MongoDB is only installed as a dependency on Travis, it's also not installed when running the unit tests using python setup.py test.

They are more like optional tests, so the two test cases should be moved into the mongodb and memcache tests respectively, and skipped if the dependency is not installed

**Comment by m-vdb** [comment]
@ask I [mocked the tests](https://github.com/m-vdb/celery/blob/mvdb/backend-as-uri-method/celery/tests/backends/test_cache.py#L37-L41) for memcached so that it doesn't raise when memcached is not installed. Should I removed that too? Is this safe enough, do you have tests suites that run only mongodb / memcached / other 3rd parties plugin-like code? Because it would be nice to avoid a possible regression for those :)

**Comment by m-vdb** [comment]
what do you think about my comment @ask ?

**Comment by ask** [comment]
Merged into master and 3.1!

@m-vdb The Travis CI build runs mongodb backend tests, and I just changed that to include cache backend tests.

Could you also add your name to https://github.com/celery/celery/edit/master/CONTRIBUTORS.txt ?

**Comment by m-vdb** [comment]
thanks @ask :) So cool. Are you planning to do a PyPi release?

**Comment by m-vdb** [comment]
https://github.com/celery/celery/pull/3083

**Pull request closed without merging**